GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/Utilities/ProcessRunner.m

```diff
@@ -0,0 +1,59 @@
+#import "ProcessRunner.h"
+
+@interface ProcessRunner()
+
+@property (strong) ProcessFactory processFactory;
+
+@end
+
+@implementation ProcessRunner
+
+- (id) initWithProcessFactory:(ProcessFactory) processFactory
+{
+    if (self = [super init])
+    {
+        self.processFactory = processFactory;
+    }
+    
+    return self;
+}
+
+- (BOOL) tryRunExecutable:(NSURL *) path
+    args:(NSArray *) args
+    output:(NSString **) output
+    error:(NSError **) error
+{
+    NSTask *task = self.processFactory();
+    NSPipe *taskOut = [NSPipe pipe];
+    
+    task.executableURL = path;
+    task.arguments = args;
+    task.standardOutput = taskOut;
+    task.standardError = taskOut;
+    
+    int exitCode = -1;
+    
+    if ([task launchAndReturnError:error])
+    {
+        [task waitUntilExit];
+        
+        exitCode = [task terminationStatus];
+        
+        *output = [[NSString alloc]
+            initWithData:[taskOut.fileHandleForReading availableData]
+            encoding:NSUTF8StringEncoding];
+        
+        if (0 != exitCode)
+        {
+            NSDictionary *userInfo = @{ NSLocalizedDescriptionKey : *output };
+            *error = [NSError errorWithDomain:@"ProcessRunner"
+                code:exitCode
+                userInfo:userInfo];
+            *output = nil;
```

Comment:
- Why is `output` only set if the `exitCode` is 0?
- Related, should `standardOutput` and `standardError` be written to different pipes (so that we can keep them separate)?

(I am guessing with how `tryRunExecutable` is used today maybe we don't care about keeping them separate?)

Reply:
>> Why is output only set if the exitCode is 0?

I'm trying to follow the "Try" pattern used in rest of our C# code here. If method fails it would return `true` and `output` will contain the command's output, `error` will be `NULL`. In case of failure, it returns `false`, `output` will be `NULL` and `error` will hold the command's output along with its `exit code`. I will update the method with a comment describing this behavior. Let me know if this is confusing.

>> Related, should standardOutput and standardError be written to different pipes (so that we can keep them separate)? (I am guessing with how tryRunExecutable is used today maybe we don't care about keeping them separate?)

Yes. Right now we only care about the `exit code`. If command succeeded we try to parse the output (combined `stdout` & `stderr`) for version strings, otherwise we assume the output is error and log it.